parsers/cxx/cxx_token.c

```diff
@@ -0,0 +1,165 @@
+/*
+*   Copyright (c) 2016, Szymon Tomasz Stefanek
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   This module contains functions for parsing and scanning C++ source files
+*/
+
+#include "cxx_token.h"
+
+#include "routines.h"
+#include "vstring.h"
+#include "read.h"
+
+#include "cxx_token_chain.h"
+#include "cxx_debug.h"
+#include "cxx_tag.h"
+
+#define CXX_TOKEN_POOL_MAXIMUM_SIZE 8192
+
+static CXXTokenChain * g_pTokenPool = NULL;
+static unsigned int g_uNextAnonumousIdentiferId = 0;
+
+void cxxTokenForceDestroy(CXXToken * t);
+
+void cxxTokenAPIInit(void)
+{
+	g_pTokenPool = cxxTokenChainCreate();
+}
+
+void cxxTokenAPINewFile(void)
+{
+	g_uNextAnonumousIdentiferId = 0;
+}
+
+void cxxTokenAPIDone(void)
+{
+	CXXToken * t;
+	
+	while((t = cxxTokenChainTakeFirst(g_pTokenPool)))
+		cxxTokenForceDestroy(t);
+
+	cxxTokenChainDestroy(g_pTokenPool);
+}
+
+CXXToken * cxxTokenCreate(void)
+{
+	CXXToken * t;
+	
+	if(g_pTokenPool->iCount > 0)
+	{
+		t = cxxTokenChainTakeFirst(g_pTokenPool);
+		CXX_DEBUG_ASSERT(t->pszWord,"The string shouldn't have been destroyed");
+	} else {
+		t = xMalloc(sizeof(CXXToken),CXXToken);
+		t->pszWord = vStringNew(); // we almost always want a string, and since this token is being reused..well.. we always want it
+	}
+
```

Comment:
How about filling t with 0 with memset?
When we add a new field to CXXToken, memset will helps us.
Of, course, pszWord must be restored after filling.